### DISCUSSION TITLE
Optimizing actions

### DISCUSSION THREAD

**Discussion by kudaba** [community/community#25539] [Actions]
<p>So during the whole setup of my actions, I had no idea there were any limits to overall usage since the documentation only talked about concurrency limits, not total time limits. I got surprise billed since the default billing option is unlimited. While annoying, that’s fine, it was only a few bucks.</p>
<p>My real questions is: what can we do to optimize our build times? I’m definitely going to remove some of the extraneous platforms now that I know about it, however there’s a few things going on that are outside my ability to control:</p>
<ol>
<li>
<p>Windows time is charged double, but it’s by far the longest to compile.</p>
</li>
<li>
<p>I have to use devenv to build and it has very expensive first time startup costs, line 28-29 is when devenv is invoked, it took 9 minutes just to start.</p>
</li>
</ol>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/5/55e4e89301dbdd727309c75e4f0afd62290e5871.png" data-download-href="/uploads/short-url/cfR1bHr54sy6RLbIPdmhb9NGMIF.png?dl=1" title="devenv start.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/22896740/181091756-bfebe620-35ff-47c1-8970-b70af8c473f0.png" alt="devenv start.png" data-base62-sha1="cfR1bHr54sy6RLbIPdmhb9NGMIF" width="690" height="99" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/5/55e4e89301dbdd727309c75e4f0afd62290e5871_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">devenv start.png</span><span class="informations">969×140 7.18 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<ol start="3">
<li>
<p>How can we deal with dependencies on external SDK? I use Vulkan which takes 2 minutes each run to install.</p>
</li>
<li>
<p>When trying to make any changes to CI, there’s alot of trial and error involved when pushing changes to the yml file. I probably blew 90% of my first months budget just getting things setup.</p>
</li>
</ol>
<p>I think for the short term I’ll collapse seperate builds (debug,release,retail) into a single run so I the vulkan and deven costs don’t hit me too hard, but what can we do about some of these redundant, fixed costs in the long term? Is there any way to iterate on the build config without eating too much time other than disabling everything while iterating?</p>

**Comment by jherico** [comment, marked as answer]
<p>For 1, you just have to research things that generally help build speed locally.  If you’re building C++/C consider looking into precompiled headers, and/or investigate building with Clang (in my experience the Clang compiler ends up being much faster than MSVC)</p>
<p>For 2, you should try to determine what devenv is doing and where it’s putting artifacts that incur the startup cost, and then modify your workflow to try to cache those locations.  Unfotunately the caching right now is very limited (2 GB per repository, which is easy to blow past for a Windows C++/C build toolchain).  On the other hand if this startup time is orthogonal to your actual project configuration, maybe file an issue here:  <a href="https://github.com/actions/virtual-environments" rel="nofollow noopener">https://github.com/actions/virtual-environments</a> suggesting that they pre-run the command on the Windows build host images.  </p>
<p>For 3, follow this issue:  <a href="https://github.com/actions/virtual-environments/issues/18" rel="nofollow noopener">https://github.com/actions/virtual-environments/issues/18</a>  It really seems to me that Vulkan should be part of the default build image.</p>
<p>For 4, you might try looking into the self-hosting functionality.  You could self-host on your own development machine and then do the bulk of the workflow development without incurring any cost, and only switch back to the Github Actions VMs at the end to ensure that it still works or debug any remaining issues.  Unfortunately Github Actions doesn’t really provide any means of reproducing the build environment / installed software that they setup, so that can be a source of bugs trying to run the same workflow between a local machine and the GA VMs.  </p>

**Comment by kudaba** [comment]
<p>Thanks for all the responses.</p>
<ol>
<li>
<p>This was mostly just a comment about msvc being slower but then the time is charged more. Now that I know about the costs I’m going to reduce my matrix to do multiple compiles in one run to reduce setup cost, and not target multiple versions of OSes.</p>
</li>
<li>
<p>Thanks for the suggession I opened an issues about it there: <a href="https://github.com/actions/virtual-environments/issues/128" rel="nofollow noopener">https://github.com/actions/virtual-environments/issues/128</a></p>
</li>
<li>
<p>Thanks again, I followed and updated with my install script as an example</p>
</li>
<li>
<p>Thats a great idea. Maybe it should be pushed more when introducing people to actions, I didn’t know I could just hook up my own pc for testing.</p>
</li>
</ol>
